### PR TITLE
Add support for vaadin web framework

### PR DESCRIPTION
**Pull request by laurit**
Vaadin relies heavily on javascript which results in completely undescriptive traces as almost every user action results in a javascript rpc call. This pr attempts to improve the situation in following ways
- set server span name to currently displayed route
- if route is not available set server span name to name of request handler that accepted the request
- add spans describing rpc calls

### CODE REVIEW

File: instrumentation/vaadin-14.2/vaadin-testing/src/main/groovy/test/vaadin/AbstractVaadinTest.groovy [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import org.testcontainers.Testcontainers
+import org.testcontainers.containers.BrowserWebDriverContainer
+import org.testcontainers.containers.output.Slf4jLogConsumer
+import spock.lang.Shared
+
+abstract class AbstractVaadinTest extends AgentInstrumentationSpecification implements HttpServerTestTrait<ConfigurableApplicationContext> {
+  private static final Logger logger = LoggerFactory.getLogger(AbstractVaadinTest)
+
+  @Shared
+  BrowserWebDriverContainer<?> chrome
```
Comment (trask) [inline instrumentation/vaadin-14.2/vaadin-testing/src/main/groovy/test/vaadin/AbstractVaadinTest.groovy:31]: 👍

File: settings.gradle [hunk 2]
```diff
@@ -250,6 +250,9 @@ include ':instrumentation:struts-2.3:javaagent'
 include ':instrumentation:tomcat-7.0:javaagent'
 include ':instrumentation:twilio-6.6:javaagent'
 include ':instrumentation:undertow-1.4:javaagent'
+include ':instrumentation:vaadin-14.2:javaagent'
+include ':instrumentation:vaadin-14.2:vaadin-14.2-testing'
+include ':instrumentation:vaadin-14.2:vaadin-testing'
```
Comment (trask) [inline settings.gradle:255]: is it possible to use `testSets` to be more consistent with how other instrumentations do multi-version testing?

File: instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        // append event type to make span name more descriptive
+        spanName += "/" + eventType;
+      }
+    }
+    return super.startSpan(spanName);
+  }
+
+  @Override
+  protected String getInstrumentationName() {
+    return "io.opentelemetry.javaagent.vaadin";
```
Comment (trask) [inline instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java:144]: we changed convention recently (#2433)
```suggestion
    return "io.opentelemetry.javaagent.vaadin-14.2";
```

File: instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinInstrumentationModule.java [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+        context = tracer().startVaadinServiceSpan(vaadinService, method);
+        scope = context.makeCurrent();
+      }
+
+      @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+      public static void onExit(
+          @Advice.Thrown Throwable throwable,
+          @Advice.Local("otelContext") Context context,
+          @Advice.Local("otelScope") Scope scope) {
+        if (scope == null) {
```
Comment (iNikem) [inline instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinInstrumentationModule.java:95]: `scope` cannot really be null here, can it?

File: instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+
+  private static final VaadinTracer TRACER = new VaadinTracer();
+
+  public static VaadinTracer tracer() {
+    return TRACER;
+  }
+
+  private VaadinTracer() {
+    super(GlobalOpenTelemetry.get());
+  }
```
Comment (iNikem) [inline instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java:37]: @anuraaga What is our current preferred form of tracer constructors?

File: instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+    String spanName = spanNameForMethod(requestHandler.getClass(), method);
+    VaadinServiceContext vaadinServiceContext = current.get(SERVICE_CONTEXT_KEY);
+    if (vaadinServiceContext != null && !vaadinServiceContext.isRequestHandled()) {
+      Span span = ServerSpan.fromContextOrNull(current);
+      if (span != null) {
+        // set server span name to request handler name
+        // we don't really know whether this request handler is going to be the one
+        // that process the request, if it isn't then next handler will also update
+        // server span name
```
Comment (iNikem) [inline instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java:84]: Do handlers call each other (and thus they are nested) or are they called one after another?

File: instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+  private static class VaadinServiceContext {
+    final String vaadinServiceSpanName;
+    boolean requestHandled;
+
+    VaadinServiceContext(String vaadinServiceSpanName) {
+      this.vaadinServiceSpanName = vaadinServiceSpanName;
+    }
+
+    void setRequestHandled() {
```
Comment (iNikem) [inline instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java:155]: Do we need some thread-safety here?

File: instrumentation/vaadin-14.2/javaagent/vaadin-14.2-javaagent.gradle [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+  testLibrary 'com.vaadin:vaadin-spring-boot-starter:16.0.0'
+
+  testImplementation project(':instrumentation:vaadin-14.2:vaadin-testing')
+  testImplementation(project(':testing-common')) {
+    exclude(module: 'jetty-server')
+  }
+
+  testInstrumentation project(':instrumentation:servlet:servlet-3.0:javaagent')
+  testInstrumentation project(':instrumentation:servlet:servlet-common:javaagent')
```
Comment (iNikem) [inline instrumentation/vaadin-14.2/javaagent/vaadin-14.2-javaagent.gradle:37]: Please take a look that changes from https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/2609 don't mess with you

File: instrumentation/vaadin-14.2/javaagent/vaadin-14.2-javaagent.gradle [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+  testInstrumentation project(':instrumentation:tomcat-7.0:javaagent')
```
Comment (iNikem) [inline instrumentation/vaadin-14.2/javaagent/vaadin-14.2-javaagent.gradle:38]: Why jetty from `testing-common` is not good enough?

File: instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinInstrumentationModule.java (hunk 4 above)
Comment (laurit) [reply instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinInstrumentationModule.java]: Fixed.

File: instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java (hunk 6 above)
Comment (laurit) [reply instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java:85]: As far as I understand both can happen. Main request processing calls handlers sequentially, but there are some handlers that call other handlers. Our instrumentation should only care about handlers in main processing and ignore the rest.

File: instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java (hunk 7 above)
Comment (laurit) [reply instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java:156]: I'd say no, it should all happen on the same thread, though making it a bit more thread safe wouldn't hurt either.

File: instrumentation/vaadin-14.2/javaagent/vaadin-14.2-javaagent.gradle (hunk 8 above)
Comment (laurit) [reply instrumentation/vaadin-14.2/javaagent/vaadin-14.2-javaagent.gradle]: Rebased.

File: instrumentation/vaadin-14.2/javaagent/vaadin-14.2-javaagent.gradle (hunk 9 above)
Comment (laurit) [reply instrumentation/vaadin-14.2/javaagent/vaadin-14.2-javaagent.gradle:55]: `testing-common` has an ancient jetty version and `org.springframework.boot:spring-boot-starter-web` already depends on tomcat. Excluding jetty from `testing-common` is done in multiple projects, it might be better if `testing-common` didn't depend on jetty.

File: instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinInstrumentationModule.java [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+        context = tracer().startRpcInvocationHandlerSpan(rpcInvocationHandler, method, jsonObject);
+        scope = context.makeCurrent();
+      }
+
+      @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+      public static void onExit(
+          @Advice.Thrown Throwable throwable,
+          @Advice.Local("otelContext") Context context,
+          @Advice.Local("otelScope") Scope scope) {
+        if (scope == null) {
```
Comment (iNikem) [inline instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinInstrumentationModule.java:272]: scope cannot be null here

File: instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinInstrumentationModule.java [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+        context = tracer().startClientCallableSpan(componentClass, methodName);
+        scope = context.makeCurrent();
+      }
+
+      @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+      public static void onExit(
+          @Advice.Thrown Throwable throwable,
+          @Advice.Local("otelContext") Context context,
+          @Advice.Local("otelScope") Scope scope) {
+        if (scope == null) {
```
Comment (iNikem) [inline instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinInstrumentationModule.java:320]: same here

File: instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java (hunk 5 above)
Comment (anuraaga) [reply instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java:38]: Ah sorry for late reply. I don't think it matters too much for Javaagent instrumentation so this seems ok.

**Review by iNikem — APPROVED** [review]

File: instrumentation/vaadin-14.2/testing/src/main/groovy/test/vaadin/AbstractVaadin14Test.groovy [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+    return handlers
+  }
+
+  @Override
+  void assertFirstRequest() {
+    assertTraces(VAADIN_14_4 ? 5 : 4) {
+      def handlers = getRequestHandlers("BootstrapHandler")
+      trace(0, 3 + handlers.size()) {
+        basicSpan(it, 0, getContextPath() + "/main", null)
```
Comment (mateuszrzeszutek) [inline instrumentation/vaadin-14.2/testing/src/main/groovy/test/vaadin/AbstractVaadin14Test.groovy:35]: I've just realized that `basicSpan()` does not check span kind at all...
`span(0)` should always be a SERVER span, right? Can you add a helper method in the base test class that does that? (I assume changing `basicSpan()` should be left for another PR)

File: instrumentation/vaadin-14.2/testing/src/main/groovy/test/vaadin/AbstractVaadin14Test.groovy [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+      }
+      trace(1, 2) {
+        basicSpan(it, 0, getContextPath() + "/*", null)
+        basicSpan(it, 1, "ApplicationDispatcher.forward", span(0))
+      }
+      trace(2, 2) {
+        basicSpan(it, 0, getContextPath() + "/*", null)
+        basicSpan(it, 1, "ApplicationDispatcher.forward", span(0))
+      }
+      trace(3, 2) {
```
Comment (mateuszrzeszutek) [inline instrumentation/vaadin-14.2/testing/src/main/groovy/test/vaadin/AbstractVaadin14Test.groovy:52]: Are those other traces significant at all? If it's possible, can you leave a comment explaining what they're representing? (ex. `/favicon.ico`)

File: instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+      RpcInvocationHandler rpcInvocationHandler, Method method, JsonObject jsonObject) {
+    String spanName = spanNameForMethod(rpcInvocationHandler.getClass(), method);
+    if ("event".equals(rpcInvocationHandler.getRpcType())) {
+      String eventType = jsonObject.getString("event");
+      if (eventType != null) {
+        // append event type to make span name more descriptive
+        spanName += "/" + eventType;
+      }
+    }
+    return super.startSpan(spanName);
```
[the remark this answers is not in the document]
Comment (laurit) [reply instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java:140]: It is on the receiving end of the rpc.

File: instrumentation/vaadin-14.2/testing/src/main/groovy/test/vaadin/AbstractVaadinTest.groovy [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+  abstract void assertFirstRequest()
+
+  abstract void assertButtonClick()
+
+  def "test vaadin"() {
+    setup:
+    def driver = getWebDriver()
+    waitForStart(driver)
+
+    driver.manage().timeouts().implicitlyWait(30, TimeUnit.SECONDS)
```
[the remark this answers is not in the document]
Comment (laurit) [reply instrumentation/vaadin-14.2/testing/src/main/groovy/test/vaadin/AbstractVaadinTest.groovy:122]: Moved. If by line below you mean `driver.get(address.resolve("main").toString())` then it is needed, `waitForStart` fetches the same page and after it is loaded clears traces, so need to fetch it again. `waitForStart` clears traces because there is some background activity during the first page load that creates an unpredictable amount of traces.

[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply instrumentation/vaadin-14.2/testing/src/main/groovy/test/vaadin/AbstractVaadinTest.groovy:122]: Can you add a comment explaining why it's there?

File: instrumentation/vaadin-14.2/testing/src/main/groovy/test/vaadin/AbstractVaadin14Test.groovy (hunk 13 above)
Comment (laurit) [reply instrumentation/vaadin-14.2/testing/src/main/groovy/test/vaadin/AbstractVaadin14Test.groovy:53]: These aren't significant, they are for javascript files. Comment added.

File: instrumentation/vaadin-14.2/testing/src/main/groovy/test/vaadin/AbstractVaadin14Test.groovy (hunk 12 above)
Comment (laurit) [reply instrumentation/vaadin-14.2/testing/src/main/groovy/test/vaadin/AbstractVaadin14Test.groovy]: Added server span assertion.

File: instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinInstrumentationModule.java (hunk 10 above)
Comment (laurit) [reply instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinInstrumentationModule.java]: Null check removed.

File: settings.gradle (hunk 2 above)
Comment (laurit) [reply settings.gradle]: Done.

**Review by mateuszrzeszutek — APPROVED** [review]
